Add sidecar config for SSL to allow for secure sidecar communication

paulben opened the pull request:
Allow ssl for: inbound to nginx, outbound from nginx to service, and healthchecks.

zcahana committed:
add k8s.io/client-go dependency

This actually adds a lot of baggage into our top-level /vendor and glide.lock.
The reason is that k8s client itself vendors a lot of packages, and glide does flattening of the vendor tree.

zcahana committed:
reimplement kubernetes adapter using the official go client

This includes:
- Use watch notifications instead of polling (via the built-in cache tool
- Support fetching credentials from service account ("InClusterConfig")

zcahana committed:
add .dockerignore

zcahana committed:
add a StringSet datastructure type

paulben committed:
Initial sidecar ssl commit

paulben committed:
Initial sidecar ssl commit

rshriram commented on testing/apps/details/Dockerfile.sidecar:
```diff
@@ -12,11 +12,11 @@
 #   limitations under the License.
 
-FROM amalgam8/a8-sidecar:alpine
+FROM amalgam8/a8-sidecar:latest
```
Why this change? Changes to this file seem unnecessary

paulben replied:
Because otherwise the sidecar image doesn't pick up my go code changes, it pulls the alpine image.

paulben replied:
I backed these changes out and tested again and these are not needed, I'll revert them in the next commit.

rshriram commented on testing/apps/productpage/Dockerfile.sidecar:
```diff
@@ -12,11 +12,11 @@
 #   limitations under the License.
 
-FROM amalgam8/a8-sidecar:alpine
+FROM amalgam8/a8-sidecar:latest
```
same as above

rshriram commented on testing/apps/ratings/Dockerfile.sidecar:
```diff
@@ -12,11 +12,11 @@
 #   limitations under the License.
 
-FROM amalgam8/a8-sidecar:alpine
+FROM amalgam8/a8-sidecar:latest
```
same as above

rshriram commented on testing/apps/reviews/Dockerfile.sidecar:
```diff
@@ -12,7 +12,7 @@
 #   limitations under the License.
 
-FROM amalgam8/a8-sidecar:alpine
+FROM amalgam8/a8-sidecar:latest
```
same as above

rshriram commented on sidecar/register/healthcheck/http.go:
```diff
@@ -45,13 +47,33 @@ func NewHTTP(conf config.HealthCheck) (Check, error) {
-		},
-		method: conf.Method,
+		client: client,
+        method: conf.Method,
```
`make precommit` should resolve these bad indents..

rshriram commented on sidecar/config/flags.go:
```diff
@@ -22,31 +22,36 @@ import (
+	configFlag                  = "config"
+	registerFlag                = "register"
+	proxyFlag                   = "proxy"
+	proxyTypeFlag               = "proxy_type"
```
may be rename to `proxy_tls_on` flag?

paulben replied:
The intent was to be consistent with endpoint:

```go
endpointTypeFlag = "endpoint_type"
```

If that consistency is not what we want then I can easily enough make it a bool flag.

rshriram replied:
So, we have to consider the case that the proxy could be just doing TCP proxying and no HTTP. In which case, TLS still applies. Which is why I suggested TLS being a boolean attribute.

rshriram commented on sidecar/config/default.go:
```diff
@@ -22,10 +22,18 @@ var DefaultConfig = Config{
 	DNS:      false,
 
+	ProxyConfig: ProxyConfig{
+		Type         : "http",
```
May be rename to `TLS : false` ?

rshriram commented on sidecar/config/default.go:
```diff
@@ -68,6 +76,8 @@ var DefaultConfig = Config{
 
 	HealthChecks: nil,
 
+	HealthChecksCACertPath: "",
```
Can be moved under HealthChecks config block.

rshriram commented on sidecar/config/config.go:
```diff
@@ -121,15 +124,25 @@ const (
 
+// Proxy configuration.
+type ProxyConfig struct {
+	Type         string `yaml:"type"`
```
rename to 
```go
Tls bool `yaml: "tls_on"
``` 
or something

rshriram commented on sidecar/config/config.go:
```diff
@@ -146,6 +159,9 @@ type Config struct {
 	Supervise bool `yaml:"supervise"`
 
 	HealthChecks []HealthCheck `yaml:"healthchecks"`
+	HealthChecksCACertPath string `yaml:"healthchecks_ca_cert_path"`
```
This should go into HealthChecks struct.

paulben replied:
There is CACertPath in HealthChecks struct. This one is intended for ease of config via envir var and cmd line. The intent is to be similar to other health check stuff - you can specify the URL easily with A8_HEALTHCHECKS but for more detailed config have to use yaml. 

If I remove it we're basically saying health check tls must be configured in yaml...right?

paulben replied:
Specifically HealthChecksCACertPath allowed me to test with this instead of yaml:

``` yaml
helloworld-v1:
  image: amalgam8/a8-examples-helloworld-sidecar:v1
  volumes:
   - 'certs:/opt/my/certs'
  environment:
    - A8_REGISTRY_URL=http://registry:8080
    - A8_SERVICE=helloworld:v1
    - A8_REGISTER=true
    - A8_ENDPOINT_PORT=5000
    - A8_ENDPOINT_TYPE=https
    - A8_HEALTHCHECKS=https://localhost:5000/health
    - A8_HEALTHCHECKS_CA_CERT_PATH=/opt/my/certs/ca_trust.pem
    - A8_LOG_LEVEL=debug
  external_links:
    - registry
```

rshriram replied:
I was really trying to cut down on the number of environment variables needed. We are trying to force people to using the YAML config for most part instead of creating umpteen number of CLI flags, as it tends to get more confusing.

In this case, I would say, yes. We are probably forcing people to use YAML config if they want to use HTTPs health checks reliably.

@zcahana @gicaceres @ijsnellf any thoughts?

ijsnellf replied:
I'm alright either way. Removing `A8_HEALTHCHECKS_CA_CERT_PATH` does make things simpler, but I can see why @paulben added it for convenience.

rshriram commented on sidecar/config/config.go:
```diff
@@ -426,3 +450,66 @@ func parseServiceNameAndTags(service string) (name string, tags []string) {
+		return err
+	}
+	return err
+}
```
This entire piece of code can go into a separate file proxyconfig.go or something, so that we can use it as a generic interface to generate configuration for any underlying proxy implementation (nginx/envoy/etc)

paulben replied:
Will do.

rshriram commented on sidecar/config/config_test.go:
```diff
@@ -360,12 +400,14 @@ log_level: debug
 			Expect(c.HealthChecks[1].Timeout).To(Equal(time.Duration(3) * time.Second))
 			Expect(c.HealthChecks[1].Method).To(Equal("POST"))
 			Expect(c.HealthChecks[1].Code).To(Equal(201))
+			Expect(c.HealthChecks[1].CACertPath).To(Equal("/etc/certs/ca2.pem"))
```
Thanks for adding the test cases for config. It would be nice to test the code in the healthcheck logic that instantiates a TLS client object for health checking.

rshriram reviewed (changes requested):
Thanks for this PR. My comments are inline.

GregHanson committed:
sidecar PID1 calls itself again with same args and env vars and does zombie reaping.  child sidecar does not perform zombie reaping

GregHanson committed:
ran make precommit

rshriram committed:
Merge pull request #421 from amalgam8/zombie_reaping_v2

Zombie reaping without dumb-init

zcahana committed:
moved Min/Max() functions 'amalgam8/pkg/math' package

zcahana committed:
kubernetes adapter fixes and improvements #1

- Switch to use a lower-level RESTClient
- Bugfix in bookkeeping of the podServices data structure
- Index pods by their names rather than by UIDs
- Fetch cached pods/endpoints by key rather than by linear search
- Comments fix

zcahana committed:
kubernetes adapter fixes and improvements #2

- Use a regular Mutex instead of RWMutex (single reader expected)
- Use finer-grain locking
- Use a single-worker workqueue to process watch events
- Avoid re-processing endpoints/pod updates when resource version hasn't changed (on full resync)

ijsnellf commented on sidecar/config/config.go:
```diff
@@ -426,3 +450,66 @@ func parseServiceNameAndTags(service string) (name string, tags []string) {
 	return
 }
+
+type A8ServicesNginxConfData struct {
```
Let's avoid prefixing structs/consts/variables with `A8` anywhere since this is the A8 code base. A8 is implicit.

This comment applies to the entire commit.

ijsnellf commented on sidecar/config/config.go:
```diff
@@ -426,3 +450,66 @@ func parseServiceNameAndTags(service string) (name string, tags []string) {
+
+func (c *Config) GenNginxA8ServicesConf() error {
+	var a8ServicesData *A8ServicesNginxConfData
+	logrus.Debug("GenProxyConfig " + strconv.FormatBool(c.Proxy) + " " + c.ProxyConfig.Type)
```
Alternate syntax that is a bit easier to read:
`logrus.Debugf("GenProxyConfig %v %v", c.Proxy, c.ProxyConfig.Type)`

ijsnellf commented on sidecar/nginx/conf/amalgam8-services.conf.tmpl:
```diff
@@ -0,0 +1,33 @@
+server {
+       listen 6379{{.SSLFlag}};
+
+       ## Comment the previous line and enable the latter two if you are using the sidecar as a gateway
```
We should probably remove these comments, since they imply you can manually modify this configuration file.

ijsnellf commented on sidecar/nginx/conf/amalgam8-services.conf.tmpl:
```diff
@@ -0,0 +1,33 @@
+include /etc/nginx/amalgam8-access-logging.conf;
```
Add some sort of "Automatically generated file --don't modify" warning here?

ijsnellf commented on sidecar/config/config.go:
```diff
@@ -146,6 +159,9 @@ type Config struct {
 	HealthChecks []HealthCheck `yaml:"healthchecks"`
+	HealthChecksCACertPath string `yaml:"healthchecks_ca_cert_path"`
+
+	ProxyConfig ProxyConfig  `yaml:"proxyconfig"`
```
"proxy_config"

ijsnellf reviewed (changes requested):
Looks pretty good overall --a few comments about structure and style.

zcahana committed:
kubernetes adapter: construct a kubernetes.Interface go get a properly configured RESTClient

zcahana committed:
Shrink released binaries using stripping and UPX compression (#418)

* Shrink released binaries using stripping and UPX compression

* Removed --ultra-brute UPX flag

* Add linker flags to enable static linking

* Removed UPX compression from Makefile release targets

zcahana committed:
Merge branch 'master' into k8s_discovery_rework

paulben committed:
code review comments

paulben committed:
Initial commit, code review comments, several merges, squash commits

paulben committed:
Merge branch 'sidecar-ssl' of https://github.com/paulben/amalgam8 into sidecar-ssl

paulben committed:
Oops...need to change test dockerfiles' FROM image tag back to alpine

GregHanson committed:
Zombie Cleanup Re-Org (#423)

Implementing code review comments for issue #422

zcahana committed:
move 'registry/api' types to 'pkg/api' package

zcahana committed:
rename 'sidecar/api' package to 'sidecar/debug'

zcahana committed:
move 'registry/adapters' packages to 'pkg/adapters/registry'

zcahana committed:
Merge branch 'master' into k8s_discovery_rework

Conflicts:
	pkg/adapters/registry/kubernetes/client.go
	pkg/adapters/registry/kubernetes/kubernetes.go
	pkg/adapters/registry/kubernetes/types.go

rshriram committed:
Merge pull request #419 from amalgam8/k8s_discovery_rework

Kubernetes discovery adapter rework

zcahana committed:
rename registry adapters to discovery adapters

zcahana committed:
fix README.md links to registry/controller docs

rshriram committed:
Merge pull request #427 from amalgam8/readme_links_fix

Fix README.md links to registry/controller docs

paulben committed:
Run make precommit on proxy_config.go

zachidan committed:
Controller API package reorg (#430)

Controller API reorg

paulben committed:
Merge branch 'sidecar-ssl' of https://github.com/paulben/amalgam8 into sidecar-ssl

rshriram commented:
duplicate of #437
